From a10953e159b408939cf4015adffe15ae13df331b Mon Sep 17 00:00:00 2001
From: Sandie Cao <sandie.cao@deepcomputing.io>
Date: Fri, 1 Nov 2024 10:29:26 +0800
Subject: [PATCH 187/214] driver: usb: hub: Set GenesysLogic Hub Downstream
 Port

On Deepcomputing FML13V10, Adjust the left&up GenesysLogic Hub
Downstream port JK Level to pass USB2.0 eye diagram.

Signed-off-by: Sandie Cao <sandie.cao@deepcomputing.io>
---
 drivers/usb/core/hub.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
index bebc41308ca9..dfc4bce43ea1 100644
--- a/drivers/usb/core/hub.c
+++ b/drivers/usb/core/hub.c
@@ -2663,6 +2663,21 @@ int usb_new_device(struct usb_device *udev)
 		goto fail;
 	}
 
+	if (udev->descriptor.idVendor == USB_VENDOR_GENESYS_LOGIC) {
+		struct usb_bus *bus = udev->bus;
+
+		if (strcmp(udev->devpath, "1.1") == 0) {
+			dev_err(&udev->dev, "Found Left&Up GenesysLogic Hub 0x27!\n");
+
+			/*Set Downstream Port 1 JK level to 7
+			  Set Downstream Port 2 JK level to 2 -- default value*/
+			err = usb_control_msg_send(udev, 0, 0xe3,
+					   0x40, 0x0020,
+					   0x27,
+					   NULL, 0, 1000, GFP_KERNEL);
+		}
+	}
+
 	/* Create link files between child device and usb port device. */
 	if (udev->parent) {
 		struct usb_hub *hub = usb_hub_to_struct_hub(udev->parent);
-- 
2.51.1

